ystia/yorc
# Feature/gh 196 a4c sm7 jobs enhancements

# Pull Request description

## Description of the change

### What I did

Support new Jobs lifecycle for:
* Slurm jobs
* Kubernetes Jobs
* Pure Tosca generic implementations using scripts or ansible recipes

Automatic cancellation of a submitted job if the workflow is cancelled


### How I did it

### How to verify it

This PR requires Alien4Cloud 2.1.0-RC1 and the following [yorc plugin](https://bintray.com/ystia/yorc-a4c-plugin/download_file?file_path=snapshots%2Ffeature%2FGH-196-a4c-sm7-jobs-enhancements%2Falien4cloud-yorc-plugin-3.1.0-SNAPSHOT.zip).

There is 3 main uses-cases to verify:
* [ ] Slurm Jobs
* [ ] Kubernetes Jobs
* [ ] Pure TOSCA Jobs with a Spark sample

For each of those uses-cases we could use the same test scenario:
1. Run a Job that succeeds
2. Run a Job that fails
3. Run a long running Job and cancel its workflow to constat its automatic cancellation

#### Preparation steps

* Create a Slurm location to import samples correctly

* Git import the following

| Repository URL | Archive's folder | Branch or Tag |
| --------------------- | -------------------- | ------------------ | 
| https://github.com/ystia/yorc-a4c-plugin.git | tosca-samples | feature/GH-196-a4c-sm7-jobs-enhancements |
| https://github.com/alien4cloud/csar-public-library.git | org/alien4cloud/spark | v2.1.0-RC1 |  
| https://github.com/ystia/forge.git | org/ystia/nfs | develop |





#### Slurm Jobs

We need to test Slurm batches, Slurm Srun and Singularity.

##### So let's begin with batches

* Create an application based on the  `SbatchJobTemplate` template
* Deploy it 
* Run the `run` workflow and check logs

* [x] Workflow should run out in success.
* [x] Logs should appear in real time

Now let's do it again but when the job is submitted then use the cli to cancel the task
``` bash
yorc d info <deploymentID>
# Get the task id from the result
yorc d tasks cancel <deploymentID> <taskID>
```
* [x] Workflow should failed in cancelled state.
* [x] Slurm job should have been automatically cancelled

##### Now with srun

* Create an application based on the  `org.ystia.yorc.samples.job.srun.Topology` template
* Deploy it 
* Run the `run` workflow and check logs

* [x] Workflow should run out in success.
* [x] Logs should appear in real time

Now let's do it again but when the job is submitted then use the cli to cancel the task
``` bash
yorc d info <deploymentID>
# Get the task id from the result
yorc d tasks cancel <deploymentID> <taskID>
```
* [x] Workflow should failed in cancelled state.
* [x] Slurm job should have been automatically cancelled


##### Finally with Singularity

* Create an application based on the  `SlurmSingularityJobTemplate1` template
* Deploy it 
* Run the `run` workflow and check logs

* [x] Workflow should run out in success.
* [x] Logs should appear in real time

Now let's do it again but when the job is submitted then use the cli to cancel the task

As there is 2 jobs in this template and they run quite fast this part may be challenging but interesting.

``` bash
yorc d info <deploymentID>
# Get the task id from the result
yorc d tasks cancel <deploymentID> <taskID>
```
* [x] Workflow should failed in cancelled state.
* [x] Slurm job should have been automatically cancelled



#### Kubernetes Jobs

First create a Kuberentes location, add it both `org.alien4cloud.kubernetes.api.types.Container` and `org.alien4cloud.kubernetes.api.types.Job`

We will use 2 topologies one that fails, and another that is quite long to run to test success and cancellation.

##### Let's start with the one that fails

* Create an application based on the  `KubeJobFail` template
* Deploy it 
* Run the `run` workflow and check logs

* [x] Workflow should exit in error.
* [x] Logs should be displayed

##### Let's start with the one that succeeds

###### Normal behaviour

* Create an application based on the  `KubeJobLoop` template
* Deploy it 
* Run the `run` workflow and check logs

* [x] Workflow should run out in success.
* [x] Logs should appear in real time

###### With cancellation

Re-run the `run` workflow and then use the cli to cancel the task

``` bash
yorc d info <deploymentID>
# Get the task id from the result
yorc d tasks cancel <deploymentID> <taskID>
```
* [x] Workflow should failed in cancelled state.
* [x] K8S job should have been automatically cancelled
* [x] Job is not visible anymore on kubernetes


#### Pure TOSCA  Jobs

For this test we will use a spark application submitted using a Job component.

Use the attached 
[spark_topology.zip](https://github.com/ystia/yorc/files/2679424/spark_topology.zip) to create an application

This topology is setup to be run on our private openstack.
So we need to create an OpenStack location, we should define a **large** compute in order to have enough power to run multiple spark jobs in parallel.

###### Normal behaviour

* Create an application based on the  `SparkJobTemplate` template
* Deploy it 
* Master Web UI is available on port 8080 workers ui on port 8081
* Run the `run` workflow and check logs

* [x] Workflow should run out in success.
* [x] Logs displaying status should appear at each `run` operation

###### With cancellation

Re-run the `run` workflow and then use the cli to cancel the task

``` bash
yorc d info <deploymentID>
# Get the task id from the result
yorc d tasks cancel <deploymentID> <taskID>
```
* [x] Workflow should failed in cancelled state.
* [x] Spark job should have been automatically cancelled
* [x] Job is marked cancelled in spark UI 



### Description for the changelog

* Support Jobs lifecycle enhancements (new operations `submit`, `run`, `cancel`) ([GH-196](https://github.com/ystia/yorc/issues/196))

## Applicable Issues

#196
